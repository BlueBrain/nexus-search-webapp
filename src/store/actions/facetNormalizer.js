import { mapObj } from "../../libs/utils";
import { findWhere } from "underscore";

function mapFacets(selectedFacets) {
  try {
    let map = mapObj(selectedFacets, (obj, key) => {
      let [groupName, subgroupName] = key.split(".");
      return {
        key: groupName,
        filters: { [subgroupName]: obj }
      };
    });
    return map.reduce((memo, selectedFilter) => {
      memo[selectedFilter.key] = selectedFilter.filters;
      return memo;
    }, {});
  } catch (error) {
    console.error(error);
  }
}

export const facetNormalizer = function(response) {
  return Object.keys(response)
    .filter(key => !!response[key].doc_count)
    .sort((a, b) => response[b].doc_count - response[a].doc_count)
    .map(key => {
      let filter = response[key];
      filter.key = key;
      return filter;
    });
};

function getBuckets (filterGroup, parentKey) {
  return Object.keys(filterGroup)
  .filter(key => key !== "doc_count" && key !== "key")
  .reduce((memo, key) => {
    let subGroup = filterGroup[key];
    if (!subGroup.buckets) {
      let subBuckets = getBuckets(subGroup, key) || {};
      return memo = Object.assign(memo, subBuckets)
    }
    memo[parentKey ? parentKey + "." + key : key] = subGroup;
    return memo
  }, {});
}

function flattenFiltersBySubgroup (facets, group) {
  facets.push(Object.assign(group, getBuckets(group)));
  return facets;
}

function addSelectedToFacet (filter, selected, facetBlacklist, path) {
    let key = filter.key;

    Object.keys(filter)
      .filter(key => key !== "doc_count" && key !== "key")
      .forEach(subGroupKey => {
        let subGroup = filter[subGroupKey];
        if (facetBlacklist.indexOf(subGroupKey) >= 0) {
          delete filter[subGroupKey];
          return;
        }
        if (!subGroup.buckets) {
          return;
        }
        subGroup.buckets.map(bucket => {
          if (
            selected[key] &&
            selected[key][subGroupKey] &&
            selected[key][subGroupKey].indexOf(bucket.key) >= 0
          ) {
            bucket.selected = true;
          } else {
            bucket.selected = false;
          }
          return bucket;
        });
      });
}

export const resultsToFacetWithSelection = function(
  facetResults,
  selectedFacets = {},
  facetBlacklist = []
) {
  let selected = mapFacets(selectedFacets);
  facetResults
    .filter(filter => {
      let key = filter.key;
      return facetBlacklist.indexOf(key) < 0;
    })
    .reduce(flattenFiltersBySubgroup, [])
    .forEach(filter => addSelectedToFacet(filter, selected, facetBlacklist));
  return facetResults;
};
