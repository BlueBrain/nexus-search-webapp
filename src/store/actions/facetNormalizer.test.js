import facetNormalizer from "./facetNormalizer";
import MOCK_ES_RESPONSE from "./mock-facet-response.json";
const SIMPLE_RESPONSE = {
  strain: {
    doc_count: 3,
    "@ids": {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: "C57BL/6J", doc_count: 1 },
        { key: "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato", doc_count: 1 },
        { key: "http://purl.obolibrary.org/obo/RS_0001833", doc_count: 1 }
      ]
    },
    labels: {
      doc_count_error_upper_bound: 0,
      sum_other_doc_count: 0,
      buckets: [
        { key: "C57BL/6J", doc_count: 1 },
        { key: "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato", doc_count: 1 },
        { key: "Wistar Han", doc_count: 1 }
      ]
    }
  }
};
const NESTED_RESPONSE = {
  brainLocation: {
    doc_count: 2,
    brainRegion: {
      doc_count: 2,
      "@ids": {
        doc_count_error_upper_bound: 0,
        sum_other_doc_count: 0,
        buckets: [
          {
            key: "http://api.brain-map.org/api/v2/data/Structure/1020",
            doc_count: 1
          },
          {
            key: "http://api.brain-map.org/api/v2/data/Structure/170",
            doc_count: 1
          }
        ]
      },
      labels: {
        doc_count_error_upper_bound: 0,
        sum_other_doc_count: 0,
        buckets: [
          {
            key: "Dorsal part of the lateral geniculate complex",
            doc_count: 1
          },
          { key: "Posterior complex of the thalamus", doc_count: 1 }
        ]
      }
    }
  }
};
describe("facetNormalizer()", () => {
  it("matches the expected output of a simple response example", () => {
    expect(facetNormalizer(SIMPLE_RESPONSE)).toMatchSnapshot();
  });

  it("able to return nested responses", () => {
    expect(facetNormalizer(NESTED_RESPONSE)).toMatchSnapshot();
  });
  it("matches the expected output of a complex example", () => {
    expect(facetNormalizer(MOCK_ES_RESPONSE)).toMatchSnapshot();
  });
});
