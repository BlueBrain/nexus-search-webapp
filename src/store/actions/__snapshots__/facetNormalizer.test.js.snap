// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`facetNormalizer() able to return nested responses 1`] = `
Array [
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "http://api.brain-map.org/api/v2/data/Structure/1020",
        "value": "http://api.brain-map.org/api/v2/data/Structure/1020",
      },
      Object {
        "amount": 1,
        "label": "http://api.brain-map.org/api/v2/data/Structure/170",
        "value": "http://api.brain-map.org/api/v2/data/Structure/170",
      },
    ],
    "title": "brainLocation.brainRegion.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "Dorsal part of the lateral geniculate complex",
        "value": "Dorsal part of the lateral geniculate complex",
      },
      Object {
        "amount": 1,
        "label": "Posterior complex of the thalamus",
        "value": "Posterior complex of the thalamus",
      },
    ],
    "title": "brainLocation.brainRegion.labels",
    "total": undefined,
  },
]
`;

exports[`facetNormalizer() matches the expected output of a complex example 1`] = `
Array [
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "C57BL/6J",
        "value": "C57BL/6J",
      },
      Object {
        "amount": 1,
        "label": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
        "value": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
      },
      Object {
        "amount": 1,
        "label": "http://purl.obolibrary.org/obo/RS_0001833",
        "value": "http://purl.obolibrary.org/obo/RS_0001833",
      },
    ],
    "title": "strain.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "C57BL/6J",
        "value": "C57BL/6J",
      },
      Object {
        "amount": 1,
        "label": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
        "value": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
      },
      Object {
        "amount": 1,
        "label": "Wistar Han",
        "value": "Wistar Han",
      },
    ],
    "title": "strain.labels",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 2,
        "label": "http://purl.obolibrary.org/obo/NCBITaxon_10090",
        "value": "http://purl.obolibrary.org/obo/NCBITaxon_10090",
      },
      Object {
        "amount": 1,
        "label": "http://purl.obolibrary.org/obo/NCBITaxon_10116",
        "value": "http://purl.obolibrary.org/obo/NCBITaxon_10116",
      },
    ],
    "title": "species.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 2,
        "label": "Mus musculus",
        "value": "Mus musculus",
      },
      Object {
        "amount": 1,
        "label": "Rattus norvegicus",
        "value": "Rattus norvegicus",
      },
    ],
    "title": "species.labels",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 2,
        "label": "http://purl.obolibrary.org/obo/PATO_0000384",
        "value": "http://purl.obolibrary.org/obo/PATO_0000384",
      },
      Object {
        "amount": 1,
        "label": "http://purl.obolibrary.org/obo/PATO_0000383",
        "value": "http://purl.obolibrary.org/obo/PATO_0000383",
      },
    ],
    "title": "sex.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 2,
        "label": "male",
        "value": "male",
      },
      Object {
        "amount": 1,
        "label": "female",
        "value": "female",
      },
    ],
    "title": "sex.labels",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "http://api.brain-map.org/api/v2/data/Structure/1020",
        "value": "http://api.brain-map.org/api/v2/data/Structure/1020",
      },
      Object {
        "amount": 1,
        "label": "http://api.brain-map.org/api/v2/data/Structure/170",
        "value": "http://api.brain-map.org/api/v2/data/Structure/170",
      },
    ],
    "title": "brainLocation.brainRegion.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "Dorsal part of the lateral geniculate complex",
        "value": "Dorsal part of the lateral geniculate complex",
      },
      Object {
        "amount": 1,
        "label": "Posterior complex of the thalamus",
        "value": "Posterior complex of the thalamus",
      },
    ],
    "title": "brainLocation.brainRegion.labels",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "schema:Person",
        "value": "schema:Person",
      },
    ],
    "title": "schema:creator.@types",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 26,
        "label": "image/png",
        "value": "image/png",
      },
      Object {
        "amount": 9,
        "label": "text/plain; charset=UTF-8",
        "value": "text/plain; charset=UTF-8",
      },
      Object {
        "amount": 1,
        "label": "application/json",
        "value": "application/json",
      },
    ],
    "title": "distribution.mediaTypes",
    "total": undefined,
  },
]
`;

exports[`facetNormalizer() matches the expected output of a simple response example 1`] = `
Array [
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "C57BL/6J",
        "value": "C57BL/6J",
      },
      Object {
        "amount": 1,
        "label": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
        "value": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
      },
      Object {
        "amount": 1,
        "label": "http://purl.obolibrary.org/obo/RS_0001833",
        "value": "http://purl.obolibrary.org/obo/RS_0001833",
      },
    ],
    "title": "strain.@ids",
    "total": undefined,
  },
  Object {
    "facetOptions": Array [
      Object {
        "amount": 1,
        "label": "C57BL/6J",
        "value": "C57BL/6J",
      },
      Object {
        "amount": 1,
        "label": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
        "value": "GAD67-eGFPxPValb-CrexR26-LSL-tdTomato",
      },
      Object {
        "amount": 1,
        "label": "Wistar Han",
        "value": "Wistar Han",
      },
    ],
    "title": "strain.labels",
    "total": undefined,
  },
]
`;
